internal/queue/queue.go

```diff
@@ -46,79 +46,90 @@ type Options struct {
 	TriggerFn api.TriggerFunction
 }
 
-// Queue is responsible for managing the cron queue and triggering jobs.
 type Queue struct {
 	log logr.Logger
 
-	queue *eventsqueue.Processor[string, counter.Interface]
-	loop  *loop.Loop
+	triggerFn        api.TriggerFunction
+	schedulerBuilder *scheduler.Builder
+	client           clientapi.Interface
+	key              *key.Key
+
+	controlLoop loops.Interface[*queue.ControlEvent]
+	queue       *eventsqueue.Processor[string, counter.Interface]
+
+	readyCh chan struct{}
 }
 
 func New(opts Options) *Queue {
-	q := &Queue{log: opts.Log.WithName("queue")}
+	q := &Queue{
+		log:              opts.Log.WithName("queue"),
+		readyCh:          make(chan struct{}),
+		triggerFn:        opts.TriggerFn,
+		schedulerBuilder: opts.SchedulerBuilder,
+		client:           opts.Client,
+		key:              opts.Key,
+	}
 
 	q.queue = eventsqueue.NewProcessor[string, counter.Interface](
 		eventsqueue.Options[string, counter.Interface]{
-			ExecuteFn: func(counter counter.Interface) {
-				q.loop.Enqueue(&queue.ControlEvent{
-					Action: &queue.ControlEvent_ExecuteRequest{
-						ExecuteRequest: &queue.ExecuteRequest{
-							JobName:    counter.JobName(),
-							CounterKey: counter.Key(),
-						},
-					},
-				})
-			},
-			Clock: opts.Clock,
+			Clock:     opts.Clock,
+			ExecuteFn: q.execute,
 		},
 	)
 
-	cache := cache.New()
-
-	q.loop = loop.New(loop.Options{
-		Log:       q.log,
-		TriggerFn: opts.TriggerFn,
-		Queue:     q.queue,
-		Cache:     cache,
-		Staging: staging.New(staging.Options{
-			Queue: q.queue,
-		}),
-		LineGuide: lineguide.New(lineguide.Options{
-			Key:          opts.Key,
-			Client:       opts.Client,
-			SchedBuilder: opts.SchedulerBuilder,
-			Queue:        q.queue,
-			Cache:        cache,
-		}),
-	})
-
 	return q
 }
 
-// Run starts the cron queue.
 func (q *Queue) Run(ctx context.Context) error {
-	q.log.Info("queue is ready")
-	defer func() {
-		//nolint:errcheck
-		q.queue.Close()
-		q.log.Info("shut down queue")
+	act := actioner.New(actioner.Options{
+		Queue:        q.queue,
+		TriggerFn:    q.triggerFn,
+		ControlLoop:  &q.controlLoop,
+		SchedBuilder: q.schedulerBuilder,
+		Client:       q.client,
+		Key:          q.key,
+	})
+
+	ictx, cancel := context.WithCancel(context.Background())
+	routerLoop := router.New(router.Options{
+		Actioner: act,
+		Log:      q.log,
+		Cancel:   cancel,
+	})
+
+	q.controlLoop = control.New(control.Options{
+		Actioner: act,
+		Jobs:     routerLoop,
+	})
+
+	errCh := make(chan error, 2)
+	go func() {
```

Comment:
Shall we use the runner manager here for these 2 go routines?

Reply:
Good idea :slightly_smiling_face: